addons/controllers/clusterbootstrap_controller.go

```diff
@@ -300,79 +343,371 @@ func (r *ClusterBootstrapReconciler) createOrPatchclusterBootstrapFromTemplate(
+// mgmt cluster. The kapp-controller running on mgmt cluster reconciles the PackageInstall CR and creates kapp-controller resources
+// on remote workload cluster. This is required for a workload cluster and its corresponding package installations to be functional.
+func (r *ClusterBootstrapReconciler) createOrPatchKappPackageInstall(clusterBootstrap *runtanzuv1alpha3.ClusterBootstrap, cluster *clusterapiv1beta1.Cluster) error {
+	// Skip if the cluster object represents the management cluster
+	if _, exists := cluster.Labels[constants.ManagementClusterRoleLabel]; exists {
+		r.Log.Info(fmt.Sprintf("cluster %s/%s is management cluster, skip creating or patching the PackageInstall CR for kapp-controller", cluster.Namespace, cluster.Name))
+		return nil
+	}
 
-	secret := &corev1.Secret{}
+	// In order to create PackageInstall CR, we need to get the Package.Spec.RefName and Package.Spec.Version
+	packageRefName, packageVersion, err := util.GetPackageMetadata(r.context, r.Client, clusterBootstrap.Spec.Kapp.RefName, cluster.Namespace)
+	if packageRefName == "" || packageVersion == "" || err != nil {
+		// Package.Spec.RefName and Package.Spec.Version are required fields for Package CR. We do not expect them to be
+		// empty and error should not happen when fetching them from a Package CR.
+		r.Log.Error(err, fmt.Sprintf("unable to fetch Package.Spec.RefName or Package.Spec.Version from Package %s/%s",
+			cluster.Namespace, clusterBootstrap.Spec.Kapp.RefName))
+		return err
+	}
 
-	if pkg.ValuesFrom.SecretRef != "" {
-		key := client.ObjectKey{Namespace: namespace, Name: pkg.ValuesFrom.SecretRef}
-		if err := r.Get(r.context, key, secret); err != nil {
-			log.Error(err, "unable to fetch secret", "objectkey", key)
-			return nil, err
-		}
+	pkgi := &kapppkgiv1alpha1.PackageInstall{
+		ObjectMeta: metav1.ObjectMeta{
+			// The legacy addons controller uses <cluster name>-<addon name> convention for naming the PackageInstall CR.
+			// https://github.com/vmware-tanzu/tanzu-framework/blob/main/addons/controllers/package_reconciler.go#L195.
+			// util.GeneratePackageInstallName() follows the same pattern.
+			Name: util.GeneratePackageInstallName(cluster.Name, packageRefName),
+			// kapp-controller PackageInstall CR is installed under the same namespace as tanzuClusterBootstrap. The namespace
+			// is also the same as where the cluster belongs.
+			Namespace: cluster.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: clusterapiv1beta1.GroupVersion.String(),
+					Kind:       cluster.Kind,
+					Name:       cluster.Name,
+					UID:        cluster.UID,
+				},
+			},
+		},
 	}
 
-	if pkg.ValuesFrom.ProviderRef != nil {
-		gvr, err := r.getGVR(schema.GroupKind{Group: *pkg.ValuesFrom.ProviderRef.APIGroup, Kind: pkg.ValuesFrom.ProviderRef.Kind})
+	pkgiMutateFn := func() error {
+		// TODO: Followup on the following fields, only add them if needed.
```

Comment:
I believe we should also add a label containing the cluster-name. I can add that later

Reply:
I filed an GH issue to add all required label/annotations in one shot. I did not blindly adding them at the moment is because I don't know what are the use cases for adding them. 

Ref: https://github.com/vmware-tanzu/tanzu-framework/issues/1677